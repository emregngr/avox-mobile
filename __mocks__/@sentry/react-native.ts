import type { ComponentType } from 'react'

const mockedSentry = {
  init: jest.fn(),
  wrap: <T extends ComponentType<any>>(Component: T): T => Component,
  withScope: jest.fn((callback: (scope: any) => void) =>
    callback({
      setTag: jest.fn(),
      setLevel: jest.fn(),
      setUser: jest.fn(),
      setContext: jest.fn(),
    }),
  ),
  captureException: jest.fn(),
  captureMessage: jest.fn(),
  addBreadcrumb: jest.fn(),
  configureScope: jest.fn(),
  setUser: jest.fn(),
  setTag: jest.fn(),
  setContext: jest.fn(),
  setRelease: jest.fn(),
  setDist: jest.fn(),
  setEnvironment: jest.fn(),
  close: jest.fn(() => Promise.resolve(true)),
  flush: jest.fn(() => Promise.resolve(true)),
}

module.exports = mockedSentry

export type mockedSentryType = typeof mockedSentry
