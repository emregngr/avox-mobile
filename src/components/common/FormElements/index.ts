export * from './CheckboxField'
export * from './TextInputField'
