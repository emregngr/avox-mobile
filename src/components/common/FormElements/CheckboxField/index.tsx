import { Ionicons } from '@expo/vector-icons'
import React, { useCallback, useMemo } from 'react'
import type { Control } from 'react-hook-form'
import { Controller } from 'react-hook-form'
import { Pressable, TouchableOpacity, View } from 'react-native'

import { ThemedText } from '@/components/common/ThemedText'
import { getLocale } from '@/locales/i18next'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'

interface CheckboxFieldProps {
  control: Control<any>
  disabled?: boolean
  error?: string
  labelKey: string
  name: string
  onPressLabel?: () => void
}

export const CheckboxField = ({
  control,
  disabled = false,
  error,
  labelKey,
  name,
  onPressLabel,
}: CheckboxFieldProps) => {
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const getCheckboxColor = useCallback(
    (error: any, value: boolean) => {
      if (error) return colors?.error
      if (value) return colors?.success
      return colors?.onPrimary100
    },
    [colors],
  )

  return (
    <View>
      <View className="flex-row items-center">
        <Controller
          render={({ field: { onChange, value } }) => (
            <TouchableOpacity
              activeOpacity={0.7}
              className="p-1"
              disabled={disabled}
              hitSlop={10}
              onPress={() => onChange(!value)}
            >
              <View className="relative">
                <Ionicons color={getCheckboxColor(error, value)} name="square-outline" size={24} />
                {value ? (
                  <Ionicons
                    className="absolute top-0.5 left-0.5"
                    color={colors?.success}
                    name="checkmark-sharp"
                    size={20}
                  />
                ) : null}
              </View>
            </TouchableOpacity>
          )}
          control={control}
          name={name}
        />

        <Pressable className="flex-1 ml-3" hitSlop={10} onPress={onPressLabel}>
          <ThemedText className="underline" color="text-100" type="body1">
            {getLocale(labelKey)}
          </ThemedText>
        </Pressable>
      </View>

      {error ? (
        <ThemedText className="mt-1 ml-12" color="error" type="body2">
          {error}
        </ThemedText>
      ) : null}
    </View>
  )
}
