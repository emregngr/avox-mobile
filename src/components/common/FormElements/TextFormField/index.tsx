import { Ionicons } from '@expo/vector-icons'
import type { ReactElement, Ref } from 'react'
import React, { forwardRef, useMemo, useState } from 'react'
import type { Control, FieldValues, Path } from 'react-hook-form'
import { Controller } from 'react-hook-form'
import type { KeyboardTypeOptions } from 'react-native'
import { TextInput, TouchableOpacity, View } from 'react-native'

import ColoredClear from '@/assets/icons/coloredClear'
import { ThemedText } from '@/components/common/ThemedText'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import { cn } from '@/utils/common/cn'

type AutoCapitalizeType = 'none' | 'sentences' | 'words' | 'characters'
type ReturnKeyType = 'done' | 'go' | 'next' | 'search' | 'send'
type TextContentType =
  | 'none'
  | 'emailAddress'
  | 'password'
  | 'newPassword'
  | 'givenName'
  | 'familyName'
  | 'name'
  | 'username'
type AutoCompleteType =
  | 'off'
  | 'email'
  | 'password'
  | 'given-name'
  | 'family-name'
  | 'name'
  | 'username'
type SubmitBehavior = 'submit' | 'blurAndSubmit' | 'newline'

interface TextFormFieldProps<T extends FieldValues> {
  autoCapitalize?: AutoCapitalizeType
  autoCorrect?: boolean
  clearTextOnFocus?: boolean
  control: Control<T>
  editable?: boolean
  keyboardType?: KeyboardTypeOptions
  label: string
  name: Path<T>
  onSubmitEditing?: () => void
  placeholder: string
  returnKeyType?: ReturnKeyType
  secureTextEntry?: boolean
  showToggle?: boolean
  submitBehavior?: SubmitBehavior
  textContentType?: TextContentType
}

const TextFormFieldComponent = <T extends FieldValues>(
  {
    autoCapitalize = 'none',
    autoCorrect = false,
    clearTextOnFocus = false,
    control,
    editable = true,
    keyboardType = 'default',
    label,
    name,
    onSubmitEditing,
    placeholder,
    returnKeyType = 'next',
    secureTextEntry = false,
    showToggle = false,
    submitBehavior = 'blurAndSubmit',
    textContentType,
  }: TextFormFieldProps<T>,
  ref: Ref<TextInput>,
) => {
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const [isPasswordVisible, setIsPasswordVisible] = useState<boolean>(false)
  const [isFocused, setIsFocused] = useState<boolean>(false)

  const getAutoCompleteType = (): AutoCompleteType => {
    const fieldName = name.toLowerCase()
    if (fieldName.includes('email')) return 'email'
    if (fieldName.includes('password')) return 'password'
    if (fieldName.includes('firstname') || fieldName.includes('givenname')) return 'given-name'
    if (fieldName.includes('lastname') || fieldName.includes('familyname')) return 'family-name'
    if (fieldName.includes('name')) return 'name'
    if (fieldName.includes('username')) return 'username'
    return 'off'
  }

  const getTextContentType = (): TextContentType => {
    if (textContentType) return textContentType

    const fieldName = name.toLowerCase()
    if (fieldName.includes('email')) return 'emailAddress'
    if (fieldName.includes('password')) return secureTextEntry ? 'password' : 'none'
    if (fieldName.includes('firstname') || fieldName.includes('givenname')) return 'givenName'
    if (fieldName.includes('lastname') || fieldName.includes('familyname')) return 'familyName'
    if (fieldName.includes('name')) return 'name'
    if (fieldName.includes('username')) return 'username'
    return 'none'
  }

  return (
    <Controller
      render={({ field: { onBlur, onChange, value }, fieldState: { error } }) => (
        <View className="w-full">
          <ThemedText className="my-1.5 px-1" color="text-70" type="body1">
            {label}
          </ThemedText>
          <View
            className={cn(
              'flex-row items-center justify-center px-4 rounded-xl overflow-hidden bg-background-tertiary border',
              error ? 'border-error' : 'border-transparent',
            )}
          >
            <TextInput
              className={cn(
                'flex-1 py-3 text-text-100 placeholder:text-text-50 text-[16px] font-inter-medium',
              )}
              onBlur={() => {
                setIsFocused(false)
                onBlur()
              }}
              passwordRules={
                secureTextEntry
                  ? 'minlength: 6; required: lower; required: upper; required: digit;'
                  : undefined
              }
              allowFontScaling={false}
              autoCapitalize={autoCapitalize}
              autoComplete={getAutoCompleteType()}
              autoCorrect={autoCorrect}
              clearTextOnFocus={clearTextOnFocus}
              editable={editable}
              keyboardAppearance={selectedTheme === 'dark' ? 'dark' : 'light'}
              keyboardType={keyboardType}
              maxFontSizeMultiplier={1.0}
              onChangeText={onChange}
              onFocus={() => setIsFocused(true)}
              onSubmitEditing={onSubmitEditing}
              placeholder={placeholder}
              ref={ref}
              returnKeyType={returnKeyType}
              secureTextEntry={secureTextEntry && !isPasswordVisible}
              spellCheck={false}
              submitBehavior={submitBehavior}
              textAlignVertical="center"
              textContentType={getTextContentType()}
              underlineColorAndroid="transparent"
              value={value}
              enablesReturnKeyAutomatically
            />
            {secureTextEntry && showToggle ? (
              <TouchableOpacity
                activeOpacity={0.7}
                className={isFocused && value?.length > 0 ? 'mr-3' : ''}
                hitSlop={10}
                onPress={() => setIsPasswordVisible(prev => !prev)}
              >
                <Ionicons
                  color={colors?.onPrimary100}
                  name={isPasswordVisible ? 'eye-off' : 'eye'}
                  size={20}
                />
              </TouchableOpacity>
            ) : null}
            {isFocused && value?.length > 0 ? (
              <TouchableOpacity activeOpacity={0.7} hitSlop={10} onPress={() => onChange('')}>
                <ColoredClear
                  height={20}
                  primaryColor={colors?.background?.primary}
                  secondaryColor={colors?.onPrimary100}
                  width={20}
                />
              </TouchableOpacity>
            ) : null}
          </View>
          {error ? (
            <ThemedText className="my-1.5 px-1" color="error" type="body2">
              {error?.message}
            </ThemedText>
          ) : null}
        </View>
      )}
      control={control}
      name={name}
    />
  )
}

export const TextFormField = forwardRef(TextFormFieldComponent) as <T extends FieldValues>(
  props: TextFormFieldProps<T> & { ref?: Ref<TextInput> },
) => ReactElement
