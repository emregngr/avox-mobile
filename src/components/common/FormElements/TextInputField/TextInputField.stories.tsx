import type { Meta, StoryObj } from '@storybook/react-native'
import { useForm } from 'react-hook-form'
import { Alert, View } from 'react-native'

import { TextInputField } from '@/components/common/FormElements/TextInputField'
import { useEffect } from 'react'

interface FormData {
  email: string
  password: string
  firstName: string
  lastName: string
  phoneNumber: string
  username: string
  confirmPassword: string
  description: string
}

const meta: Meta<typeof TextInputField> = {
  title: 'Components/Forms/TextInputField',
  component: TextInputField,
  parameters: {
    docs: {
      description: {
        component: 'Text input field component with form validation support using react-hook-form.',
      },
    },
  },
  decorators: [
    Story => (
      <View className="p-4 bg-background-primary flex-1">
        <Story />
      </View>
    ),
  ],
  argTypes: {
    label: {
      control: { type: 'text' },
      description: 'Label text displayed above the input',
    },
    placeholder: {
      control: { type: 'text' },
      description: 'Placeholder text shown when input is empty',
    },
    name: {
      control: { type: 'text' },
      description: 'Form field name',
    },
    keyboardType: {
      control: { type: 'select' },
      options: ['default', 'email-address', 'numeric', 'phone-pad', 'number-pad'],
      description: 'Keyboard type to display',
    },
    autoCapitalize: {
      control: { type: 'select' },
      options: ['none', 'sentences', 'words', 'characters'],
      description: 'Auto-capitalization behavior',
    },
    returnKeyType: {
      control: { type: 'select' },
      options: ['done', 'go', 'next', 'search', 'send'],
      description: 'Return key type',
    },
    secureTextEntry: {
      control: { type: 'boolean' },
      description: 'Whether to hide text input (for passwords)',
    },
    showToggle: {
      control: { type: 'boolean' },
      description: 'Show password visibility toggle (only works with secureTextEntry)',
    },
    editable: {
      control: { type: 'boolean' },
      description: 'Whether the input is editable',
    },
    autoCorrect: {
      control: { type: 'boolean' },
      description: 'Whether to enable auto-correction',
    },
    clearTextOnFocus: {
      control: { type: 'boolean' },
      description: 'Whether to clear text when focused',
    },
    testID: {
      control: { type: 'text' },
      description: 'Test identifier',
    },
  },
}

export default meta

type Story = StoryObj<typeof TextInputField>

const FormWrapper = ({
  name = 'email',
  label = 'Email Address',
  placeholder = 'Enter your email',
  keyboardType = 'default',
  autoCapitalize = 'none',
  returnKeyType = 'next',
  secureTextEntry = false,
  showToggle = false,
  editable = true,
  autoCorrect = false,
  clearTextOnFocus = false,
  testID,
  onSubmitEditing,
  defaultValue = '',
  withError = false,
  errorMessage = '',
}: {
  name?: keyof FormData
  label?: string
  placeholder?: string
  keyboardType?: 'default' | 'email-address' | 'numeric' | 'phone-pad' | 'number-pad'
  autoCapitalize?: 'none' | 'sentences' | 'words' | 'characters'
  returnKeyType?: 'done' | 'go' | 'next' | 'search' | 'send'
  secureTextEntry?: boolean
  showToggle?: boolean
  editable?: boolean
  autoCorrect?: boolean
  clearTextOnFocus?: boolean
  testID?: string
  onSubmitEditing?: () => void
  defaultValue?: string
  withError?: boolean
  errorMessage?: string
}) => {
  const { control, setError } = useForm<FormData>({
    defaultValues: {
      email: '',
      password: '',
      firstName: '',
      lastName: '',
      phoneNumber: '',
      username: '',
      confirmPassword: '',
      description: '',
      [name]: defaultValue,
    },
  })

  useEffect(() => {
    if (withError && errorMessage) {
      setError(name as keyof FormData, {
        type: 'manual',
        message: errorMessage,
      })
    }
  }, [withError, errorMessage, name, setError])

  return (
    <TextInputField
      control={control}
      name={name}
      label={label}
      placeholder={placeholder}
      keyboardType={keyboardType}
      autoCapitalize={autoCapitalize}
      returnKeyType={returnKeyType}
      secureTextEntry={secureTextEntry}
      showToggle={showToggle}
      editable={editable}
      autoCorrect={autoCorrect}
      clearTextOnFocus={clearTextOnFocus}
      {...(testID && { testID })}
      {...(onSubmitEditing && { onSubmitEditing })}
    />
  )
}

export const Default: Story = {
  render: () => <FormWrapper />,
  parameters: {
    docs: {
      description: {
        story: 'Basic text input field with form integration.',
      },
    },
  },
}

export const Email: Story = {
  render: () => (
    <FormWrapper
      name="email"
      label="Email Address"
      placeholder="Enter your email"
      keyboardType="email-address"
      autoCapitalize="none"
    />
  ),
  parameters: {
    docs: {
      description: {
        story: 'Email input field with proper keyboard type and auto-capitalization.',
      },
    },
  },
}

export const Password: Story = {
  render: () => (
    <FormWrapper
      name="password"
      label="Password"
      placeholder="Enter your password"
      secureTextEntry={true}
      showToggle={true}
      returnKeyType="done"
    />
  ),
  parameters: {
    docs: {
      description: {
        story: 'Password input field with visibility toggle.',
      },
    },
  },
}

export const PasswordNoToggle: Story = {
  render: () => (
    <FormWrapper
      name="confirmPassword"
      label="Confirm Password"
      placeholder="Confirm your password"
      secureTextEntry={true}
      showToggle={false}
      returnKeyType="done"
    />
  ),
  parameters: {
    docs: {
      description: {
        story: 'Password input field without visibility toggle.',
      },
    },
  },
}

export const Numeric: Story = {
  render: () => (
    <FormWrapper
      name="phoneNumber"
      label="Phone Number"
      placeholder="Enter phone number"
      keyboardType="phone-pad"
      returnKeyType="done"
    />
  ),
  parameters: {
    docs: {
      description: {
        story: 'Numeric input field with phone pad keyboard.',
      },
    },
  },
}

export const WithError: Story = {
  render: () => (
    <FormWrapper
      name="email"
      label="Email Address"
      placeholder="Enter your email"
      keyboardType="email-address"
      defaultValue="invalid-email"
      withError={true}
      errorMessage="Please enter a valid email address"
    />
  ),
  parameters: {
    docs: {
      description: {
        story: 'Text input field with validation error message.',
      },
    },
  },
}

export const Disabled: Story = {
  render: () => (
    <FormWrapper
      name="username"
      label="Username (Disabled)"
      placeholder="This field is disabled"
      editable={false}
      defaultValue="john_doe"
    />
  ),
  parameters: {
    docs: {
      description: {
        story: 'Disabled text input field that cannot be edited.',
      },
    },
  },
}

export const LongText: Story = {
  render: () => (
    <FormWrapper
      name="description"
      label="Description"
      placeholder="Enter a detailed description..."
      autoCapitalize="sentences"
      returnKeyType="done"
    />
  ),
  parameters: {
    docs: {
      description: {
        story: 'Text input field for longer text with sentence capitalization.',
      },
    },
  },
}

export const WithSubmitAction: Story = {
  render: () => (
    <FormWrapper
      name="firstName"
      label="First Name"
      placeholder="Enter your first name"
      autoCapitalize="words"
      returnKeyType="next"
      onSubmitEditing={() => Alert.alert('Submit', 'Next field would be focused')}
    />
  ),
  parameters: {
    docs: {
      description: {
        story: 'Text input field with submit action on return key press.',
      },
    },
  },
}

export const MultipleInputs: Story = {
  render: () => {
    const { control } = useForm<FormData>({
      defaultValues: {
        email: '',
        password: '',
        firstName: '',
        lastName: '',
        phoneNumber: '',
        username: '',
        confirmPassword: '',
        description: '',
      },
    })

    const inputs = [
      {
        name: 'firstName' as keyof FormData,
        label: 'First Name',
        placeholder: 'Enter your first name',
        autoCapitalize: 'words' as const,
      },
      {
        name: 'lastName' as keyof FormData,
        label: 'Last Name',
        placeholder: 'Enter your last name',
        autoCapitalize: 'words' as const,
      },
      {
        name: 'email' as keyof FormData,
        label: 'Email Address',
        placeholder: 'Enter your email',
        keyboardType: 'email-address' as const,
        autoCapitalize: 'none' as const,
      },
      {
        name: 'phoneNumber' as keyof FormData,
        label: 'Phone Number',
        placeholder: 'Enter phone number',
        keyboardType: 'phone-pad' as const,
      },
      {
        name: 'password' as keyof FormData,
        label: 'Password',
        placeholder: 'Enter password',
        secureTextEntry: true,
        showToggle: true,
        returnKeyType: 'done' as const,
      },
    ]

    return (
      <View className="gap-4">
        {inputs.map((input, index) => (
          <TextInputField
            key={input.name}
            control={control}
            name={input.name}
            label={input.label}
            placeholder={input.placeholder}
            testID={`input-${index}`}
            keyboardType={input.keyboardType || 'default'}
            autoCapitalize={input.autoCapitalize || 'none'}
            returnKeyType={input.returnKeyType || 'next'}
            secureTextEntry={input.secureTextEntry || false}
            showToggle={input.showToggle || false}
          />
        ))}
      </View>
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'Multiple text input fields in a form with various configurations.',
      },
    },
  },
}

export const AllStates: Story = {
  render: () => {
    const { control, setError } = useForm({
      defaultValues: {
        normal: '',
        withValue: 'Filled value',
        error: 'invalid@email',
        disabled: 'Disabled value',
        password: '',
      },
    })

    useEffect(() => {
      setError('error', {
        type: 'manual',
        message: 'Invalid email format',
      })
    }, [setError])

    return (
      <View className="gap-4">
        <View>
          <TextInputField
            control={control}
            name="normal"
            label="Normal empty state"
            placeholder="Type here..."
          />
        </View>

        <View>
          <TextInputField
            control={control}
            name="withValue"
            label="With value"
            placeholder="This has a value"
          />
        </View>

        <View>
          <TextInputField
            control={control}
            name="error"
            label="With error state"
            placeholder="Email with error"
            keyboardType="email-address"
          />
        </View>

        <View>
          <TextInputField
            control={control}
            name="disabled"
            label="Disabled state"
            placeholder="Can't edit this"
            editable={false}
          />
        </View>

        <View>
          <TextInputField
            control={control}
            name="password"
            label="Password with toggle"
            placeholder="Enter password"
            secureTextEntry={true}
            showToggle={true}
          />
        </View>
      </View>
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'All possible states of the text input field in one view.',
      },
    },
  },
}

export const Playground: Story = {
  render: (args: any) => {
    const {
      name,
      label,
      placeholder,
      keyboardType,
      autoCapitalize,
      returnKeyType,
      secureTextEntry,
      showToggle,
      editable,
      autoCorrect,
      clearTextOnFocus,
      testID,
    } = args

    return (
      <FormWrapper
        name={name as keyof FormData}
        label={label}
        placeholder={placeholder}
        keyboardType={keyboardType}
        autoCapitalize={autoCapitalize}
        returnKeyType={returnKeyType}
        secureTextEntry={secureTextEntry}
        showToggle={showToggle}
        editable={editable}
        autoCorrect={autoCorrect}
        clearTextOnFocus={clearTextOnFocus}
        {...(testID && { testID })}
      />
    )
  },
  args: {
    name: 'email',
    label: 'Email Address',
    placeholder: 'Enter your email',
    keyboardType: 'email-address',
    autoCapitalize: 'none',
    returnKeyType: 'next',
    secureTextEntry: false,
    showToggle: false,
    editable: true,
    autoCorrect: false,
    clearTextOnFocus: false,
    testID: 'playground-input',
  },
  parameters: {
    docs: {
      description: {
        story: 'Interactive playground to test all props and configurations.',
      },
    },
  },
}
