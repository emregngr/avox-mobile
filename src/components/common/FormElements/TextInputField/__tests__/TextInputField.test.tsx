import { fireEvent, render, waitFor } from '@testing-library/react-native'
import React from 'react'
import { Controller, useForm } from 'react-hook-form'
import { Text, TouchableOpacity } from 'react-native'

import { TextInputField } from '@/components/common/FormElements/TextInputField'
import useThemeStore from '@/store/theme'

jest.mock('@/store/theme')

const mockedUseThemeStore = useThemeStore as jest.MockedFunction<typeof useThemeStore>

const MockedTextInputField = (props: any) => {
  const { control } = useForm({
    defaultValues: { email: '' },
    mode: 'onBlur',
  })
  return (
    <TextInputField
      control={control}
      label="Email"
      name="email"
      placeholder="Enter email"
      {...props}
    />
  )
}

beforeEach(() => {
  mockedUseThemeStore.mockReturnValue({
    selectedTheme: 'light',
  })
})

describe('TextInputField Component', () => {
  it('renders label and placeholder', () => {
    const { getByText, getByPlaceholderText } = render(<MockedTextInputField />)
    expect(getByText('Email')).toBeTruthy()
    expect(getByPlaceholderText('Enter email')).toBeTruthy()
  })

  it('updates value on typing', () => {
    const { getByPlaceholderText } = render(<MockedTextInputField />)
    const input = getByPlaceholderText('Enter email')
    fireEvent.changeText(input, 'test@example.com')
    expect(input.props.value).toBe('test@example.com')
  })

  it('shows and clears with close icon when focused and has text', () => {
    const { getByPlaceholderText, queryByTestId, getByTestId } = render(<MockedTextInputField />)
    const input = getByPlaceholderText('Enter email')

    expect(queryByTestId('clear-button')).toBeNull()

    fireEvent.changeText(input, 'hello')
    fireEvent(input, 'focus')

    const clearButton = getByTestId('clear-button')
    expect(clearButton).toBeTruthy()

    fireEvent.press(clearButton)
    expect(input.props.value).toBe('')
  })

  it('toggles password visibility when secureTextEntry + showToggle', () => {
    const { getByPlaceholderText, getByTestId } = render(
      <MockedTextInputField
        name="password"
        placeholder="Enter password"
        secureTextEntry
        showToggle
      />,
    )
    const input = getByPlaceholderText('Enter password')
    const toggle = getByTestId('password-toggle')

    expect(input.props.secureTextEntry).toBe(true)

    fireEvent.press(toggle)
    expect(input.props.secureTextEntry).toBe(false)

    fireEvent.press(toggle)
    expect(input.props.secureTextEntry).toBe(true)
  })

  it('renders error message when error exists', async () => {
    const TestComponent = () => {
      const { control } = useForm({
        defaultValues: { email: '' },
        mode: 'onBlur',
      })

      return (
        <Controller
          render={() => (
            <TextInputField
              control={control}
              label="Email"
              name="email"
              placeholder="Enter email"
            />
          )}
          control={control}
          name="email"
          rules={{ required: 'This field is required' }}
        />
      )
    }

    const { getByText, getByPlaceholderText } = render(<TestComponent />)

    const input = getByPlaceholderText('Enter email')

    fireEvent.changeText(input, 'test')
    fireEvent.changeText(input, '')
    fireEvent(input, 'blur')

    await waitFor(() => {
      expect(getByText('This field is required')).toBeTruthy()
    })
  })

  it('renders error message after form submission attempt', async () => {
    const TestComponent = () => {
      const methods = useForm({
        defaultValues: { email: '' },
      })

      const onSubmit = () => {}

      return (
        <>
          <Controller
            render={() => (
              <TextInputField
                control={methods.control}
                label="Email"
                name="email"
                placeholder="Enter email"
              />
            )}
            control={methods.control}
            name="email"
            rules={{ required: 'This field is required' }}
          />
          <TouchableOpacity onPress={methods.handleSubmit(onSubmit)} testID="submit-button">
            <Text>Submit</Text>
          </TouchableOpacity>
        </>
      )
    }

    const { getByText, getByTestId } = render(<TestComponent />)

    const submitButton = getByTestId('submit-button')
    fireEvent.press(submitButton)

    await waitFor(() => {
      expect(getByText('This field is required')).toBeTruthy()
    })
  })
})

describe('TextInputField Component Snapshot', () => {
  it('should render the TextInputField Component successfully', () => {
    const { toJSON } = render(<MockedTextInputField />)

    expect(toJSON()).toMatchSnapshot()
  })
})
