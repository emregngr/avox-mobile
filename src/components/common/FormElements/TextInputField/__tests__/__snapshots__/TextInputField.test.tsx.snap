// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TextInputField Component Snapshot should render the TextInputField Component successfully 1`] = `
<View
  className="w-full"
>
  <Text
    allowFontScaling={false}
    className="text-[16px] leading-[19px] font-inter-medium text-text-70 my-1.5 mx-1"
    maxFontSizeMultiplier={1}
  >
    Email
  </Text>
  <View
    className="flex-row items-center px-4 rounded-xl overflow-hidden bg-background-tertiary border border-transparent"
  >
    <TextInput
      allowFontScaling={false}
      autoCapitalize="none"
      autoComplete="email"
      autoCorrect={false}
      className="flex-1 py-3 text-text-100 placeholder:text-text-50 text-[16px] font-inter-medium"
      clearTextOnFocus={false}
      editable={true}
      enablesReturnKeyAutomatically={true}
      keyboardAppearance="light"
      keyboardType="default"
      maxFontSizeMultiplier={1}
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      placeholder="Enter email"
      returnKeyType="next"
      secureTextEntry={false}
      spellCheck={false}
      submitBehavior="blurAndSubmit"
      testID="textinput"
      textAlignVertical="center"
      textContentType="emailAddress"
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
</View>
`;
